### PR TITLE
(Another) fixed reading archived data

### PR DESCRIPTION
**Pull request by ibuclaw**
Similar to #10 but differs in the following ways:
- Don't blindly aggregate to the largest timestep in the range read.  I found this to be a massive problem when you are pulling metrics from nodes that have been recording for different amounts of time  (ie: Request 3 months, but one node only has data for 3 days)

What I found better was to start at the default timeStep, then aggregate as needed when you reach some magic high number.  This ended up being the number of retentions held for the timeStep you are normalizing to.  One litmus test I used was: how fast is it to retrieve 3000 days of data? Is that faster or slow than 300 days? 30 days? 3 days? 3 hours?  And had comparable results at either end of the spectrum based on my retention configuration.
- Supporting other kinds of aggregation that carbon may set, defaulting to 'average'.  This matches in functionality something similar to that of Whisper.

I'd also like to find a way to re-factor the repeated parts into some re-useable function, but stopped when the number of passed parameters exceeded five.

### CODE REVIEW

File: ceres.py [hunk 1]
```diff
@@ -444,6 +445,12 @@ def read(self, fromTime, untilTime):
     sliceBoundary = None  # to know when to split up queries across slices
     resultValues = []
     earliestData = None
+    timeStep = self.timeStep
+
+    # If the metadata hold retentions/aggregation for slices, fetch them
+    retentions = dict(iter(self.metadata.get('retentions', [])))
```
Comment (mleinart) [inline ceres.py]: Is iter() needed here for any reason?

File: ceres.py [hunk 2]
```diff
@@ -444,6 +445,12 @@ def read(self, fromTime, untilTime):
...
+    method = self.metadata.get('aggregationMethod', 'average')
```
Comment (mleinart) [inline ceres.py]: Add a DEFAULT_AGGREGATION_METHOD constant at the top and use that

File: ceres.py [hunk 3]
```diff
@@ -444,6 +445,12 @@ def read(self, fromTime, untilTime):
...
+    bigger_timeSteps = sorted(filter(lambda x: x > timeStep, retentions))
```
Comment (mleinart) [inline ceres.py]: camelCase

File: ceres.py [hunk 4]
```diff
@@ -473,27 +499,60 @@ def read(self, fromTime, untilTime):
         except NoData:
           continue
 
+        # Calculate the cost for growing values, and aggregate current data if too big
+        maxCost = retentions.get(timeStep, 0)
+        missingGap = (requestUntilTime - series.endTime)
+        while len(bigger_timeSteps) and maxCost != 0:
+          newValues = int(len(series.values) * (series.timeStep / timeStep))
```
Comment (mleinart) [inline ceres.py]: Thinking aloud: are we relying on carbon's storage-schema validation to enforce retentions being evenly divisible? 
TBC, I dont think we should do any validation here but if it's not done somewhere I expect we'll eventually see some weird bug reports when people configure retentions unusually

File: ceres.py [hunk 5]
```diff
@@ -473,27 +499,60 @@ def read(self, fromTime, untilTime):
...
+    # Finally, aggregate for any left padded nulls.
+    maxCost = retentions.get(timeStep, 0)
+    missingGap = int(untilTime - fromTime)
+    if earliestData is not None:
+      missingGap = int(earliestData - fromTime)
+    while len(bigger_timeSteps) and maxCost != 0:
+      if (int(missingGap / timeStep) + len(resultValues)) <= maxCost:
+        break
+
+      new_timeStep = bigger_timeSteps.pop(0)
```
Comment (mleinart) [inline ceres.py]: camelCase

File: ceres.py [hunk 6]
```diff
@@ -819,6 +882,57 @@ class SliceDeleted(Exception):
...
+  knownValues = filter(lambda x: x is not None, values)
+  if len(knownValues) is 0:
+    return None
+  # Aggregate based on method
+  if aggregationMethod == 'average':
+    return float(sum(knownValues)) / float(len(knownValues))
+  elif aggregationMethod == 'sum':
+    return float(sum(knownValues))
+  elif aggregationMethod == 'last':
+    return knownValues[len(knownValues) - 1]
```
Comment (mleinart) [inline ceres.py]: Use knownValues[-1] instead

File: ceres.py [hunk 7]
```diff
@@ -819,6 +882,57 @@ class SliceDeleted(Exception):
...
+  elif aggregationMethod == 'max':
+    return max(knownValues)
+  elif aggregationMethod == 'min':
+    return min(knownValues)
+  else:
+    raise InvalidAggregationMethod("Unrecognized aggregation method %s" %
+                                   aggregationMethod)
+
+
+def recalculateSeries(method, values, old_timeStep, new_timeStep):
```
Comment (mleinart) [inline ceres.py]: camelCase

File: ceres.py [hunk 8]
```diff
@@ -819,6 +882,57 @@ class SliceDeleted(Exception):
...
+    for value in values:
+      newValues.extend([value] * ratio)
+    return newValues
+  # If the series is at a higher precision, aggregate to fit to a lower precision
+  elif old_timeStep < new_timeStep:
+    factor = int(new_timeStep / old_timeStep)
+    newValues = []
+    subArr = []
+    subArr_length = 0
+    for i in range(0, len(values)):
```
Comment (mleinart) [inline ceres.py]: for val in values:

File: tests/test_ceres.py [hunk 9]
```diff
@@ -867,3 +869,394 @@ def test_correct_size_written_one_point_gap(self, open_mock):
...
+    datapoints = [(600, 0), (600, 1)]
+    compacted = self.ceres_node.compact(datapoints)
+    self.assertEqual([[(600, 1.0)]], compacted)
+
+  def test_archives_compact_groups_contiguous_points(self):
+    datapoints = [(600, 0), (630, 0), (840, 0)]
+    compacted = self.ceres_node.compact(datapoints)
+    self.assertEqual([[(600, 0), (630, 0)], [(840, 0)]], compacted)
+
+  def test_archives_write_noops_if_no_datapoints(self):
```
Comment (mleinart) [inline tests/test_ceres.py]: Remove this and other write tests from CeresArchiveNodeTest and CeresArchiveNodeReadTest as they're generally not relevant. The one thing we'd possibly want tests for is something to ensure writes go to the timestep in self.timeStep and not some other series. Given https://github.com/graphite-project/ceres/blob/master/ceres.py#L521-L523 clearly skips those though, I consider that low priority

File: ceres.py (hunk 1 above)
Comment (ibuclaw) [inline ceres.py]: I thought it was needed.  Apparently not it seems.

**Comment by ibuclaw** [comment]
> Thinking aloud: are we relying on carbon's storage-schema validation to enforce retentions being evenly divisible?
TBC, I dont think we should do any validation here but if it's not done somewhere I expect we'll eventually see some weird bug reports when people configure retentions unusually

Whisper documents this: http://graphite.readthedocs.org/en/latest/whisper.html#archives-retention-and-precision

So I assume that they use similar assumptions too.  If there is a better way, I'd not object to it though.

File: ceres.py (hunk 6 above)
Comment (ibuclaw) [inline ceres.py]: Ah, coming from another language I'm used to writing ``[$ - 1]`` and immediately think of ``[-1]`` as an out of bounds index.

**Comment by mleinart** [comment]
Thanks for submitting this

I'm unsure of the up-sampling behavior here. It does address an annoying aspect of whisper's behavior when data from the current archive is requested plus a few points from the next archive (whisper flips to the next archive's precision even if there's only 1 point overlapping to it), but it kindof feels like it's making the data lie.

I'll note with both this and #10 that the original intention of Ceres by the author was for this behavior to be moved up to the display layer to be handled such that the user had visibility and control into how data is coerced and perhaps allowing a single graph to have multiple levels of precisions in it. However, the pragmatic path forward is clearly to support some whisper-like behavior and just move on.

I'd love some more feedback on the upsampling case. Example of this behavior:
Say I have retentions: 1w:60s, 1y:1h
Say then I request data from -8days to -5days (that's 2 days of data at 60s resolution, 1 day of data at 1h resolution)
Whisper's behavior will be to return the data in hourly resolution and be done with it
The proposed behavior (if I have it correct) would be to upsample the 1 day worth of hourly data by effectively applying a [keepLastValue()](http://graphite.readthedocs.org/en/latest/functions.html#graphite.render.functions.keepLastValue) and include it with the other 2 days of data at minutely resolution

**Comment by ibuclaw** [comment]
> I'll note with both this and #10 that the original intention of Ceres by the author was for this behavior to be moved up to the display layer to be handled such that the user had visibility and control into how data is coerced and perhaps allowing a single graph to have multiple levels of precisions in it.

Out of curiosity, how much work would it be for the graphite-web frontend to handle receiving a list of TimeSeriesData?  The main source of pain I'm trying to address is that ceres currently returns a single array of data points with no information other about how they relate to each other.

Given the read data:
``[489, 467, 461, 460, 468, 470]``

The most desirable output would be:
``[(fromA, untilA, 300, [489]), (fromB, untilB, 60, [467, 461, 460, 468, 470])``

> I'd love some more feedback on the upsampling case. Example of this behavior:
Say I have retentions: 1w:60s, 1y:1h
Say then I request data from -8days to -5days (that's 2 days of data at 60s resolution, 1 day of data at 1h resolution)
Whisper's behavior will be to return the data in hourly resolution and be done with it
The proposed behavior (if I have it correct) would be to upsample the 1 day worth of hourly data by effectively applying a keepLastValue() and include it with the other 2 days of data at minutely resolution

That is correct (example):
graphite-graph

I guess if upsampling is a problem, you've probably noted that it will upsample all data of a lower precision by default.  Some smarter alternatives could be:
- Start at the highest resolution of data in our slices and work our way down as data grows.
- Resample to the most common denominator.

But these might be difficult to get right IMO.  When testing #10, I found that the number of results returned when crossing slice boundaries across nodes collecting the same metric to be inconsistent because the initial probe for the ``biggest_timeStep`` would return a different result depending on:
- Does the node have data in the slice?
- Any data at all?
- Is the time delta between the two slice beginnings 10 seconds apart? 10 minutes?  (I say this, but in hindsight, it does not help you at all if the rollup script you're using doesn't use a fixed timestamp to align all slices to the same time).

The only way to get everything properly aligned was just to upsample to the highest .

There is a pragmatic vs. correctness to think about here for sure though.

**Comment by ibuclaw** [comment]
To clarify, the "upsample everything" consideration came from the fact that our data was like this:
```
node1/metric:
  1351987200@86400.slice  1422147600@3600.slice  1429926900@300.slice  1432519680@30.slice
node2/metric:
  1351987200@86400.slice  1422151200@3600.slice  1429927500@300.slice  1432519680@30.slice
```
And what I was trying to solve was the ability to retrieve a grouped set of data from any arbitrary range of time, so that all returned data have the same number datapoints that are in guaranteed to be in lock-step with each other.

We use our own roll-up scripts based on dkulikovsky 's fork of megacarbon, but it looks like using upstream megacarbon rollup will produce the same effect (https://github.com/graphite-project/carbon/blob/ccd6f6709c0fcff027685dd69959b9959cfb07ed/plugins/maintenance/rollup.py#L32)

I've made an amendment to my rollup script and now I'm seeing results that will make things  easier to work with:
```
node1/metric:
  1351987200@86400.slice  1422666000@3600.slice  1430443200@300.slice  1433035290@30.slice
node2/metric:
  1351987200@86400.slice  1422666000@3600.slice  1430443200@300.slice  1433035290@30.slice
```
Now, it is the front-end's job to make sure that this sort of data alignment is maintained.  But when it comes to upsampling/aggregation, should we take into consideration scenarioA, or scenarioB?

**Comment by ibuclaw** [comment]
The short statement is:  Graphite handles uneven datapoints in a bad way, like *really* bad.

**Comment by ibuclaw** [comment]
@mleinart - So, shall I try suggesting this in the finder in upstream graphite-web?  That way, the only change that needs to be done on the ceres side is to also return the timeStep associated with a read slice, rather than a single merged list of values.

**Comment by iain-buclaw-sociomantic** [comment]
Rebased on master and updated to what is currently running in production on my end.  I removed the logic to repeat values, so when reading across archives the lowest precision trumps all (unlike what is shown in the graph image I attached earlier this year).

**Comment by iain-buclaw-sociomantic** [comment]
And now went one step further and added a ``MAX_DATA_POINTS`` global variable, as the memory cost of holding the retentions dictionary in memory was getting a bit silly.

**Comment by deniszh** [comment]
Hello @ibuclaw @iain-buclaw-sociomantic !
Are you still using Ceres? How 'bout merge this?

**Comment by iain-buclaw-sociomantic** [comment]
We are still using ceres with this patch applied as-is.  I think will need to add a few more unittests to add coverage to the aggregationMethod code however.

In case the above commentary is not clear, this patch adds the following behaviour:
- Given retentions of `1w:60s, 1y:1h` and requesting data from -8days to -5days (that's 2 days of data at 60s resolution, 1 day of data at 1h resolution), it will now follow Whisper's behavior and return the data in hourly resolution, aggregating data at a higher precision according to `aggregationMethod`.
- On top of that, there is a `MAX_DATA_POINTS` global setting, which can be added to graphite-web (i.e: `CERES_MAX_DATA_POINTS`) that allows an upper-bound to be set on the number of data points returned to the graphite reader.

**Comment by iain-buclaw-sociomantic** [comment]
Moving this to #47.

**Pull request closed without merging**